### ISSUE TITLE
MacOS missing overrides for symbolic IO

### ISSUE THREAD

**Issue by robdockins** [GaloisInc/crucible#885]
I'm experiencing test suite failures on my local machine that do not occur on the CI runners. The overrides for, e.g., `write` are not found.  This appears to be because clang and/or libc on this system insist on adding a strange prefix to some (?) symbols.  In particular `write` is compiled into `\01_write` (yes, with a leading ASCII character with code `0x01`).  I suppose this is due to some MacOS/XCode nonsense which was added in some version bump between where the runners are and where my machine is (Version 11.5.1 (20G80)).

The obvious solution is to add overrides for both symbols, but the tricky part is understanding when/why this mangling gets done so we can figuring out which symbols need special treatment.

```
crux-llvm
  clang 11.1.0
    write-stdout.c
      z3
        loop
          clang11
            z3 4.8.10
              write-stdout
                write-stdout crux run:    OK (0.45s)
                write-stdout crux output: FAIL
                  test/Test.hs:234:
                  MISMATCH for expected (test-data/golden/stdio/write-stdout.good)
                             and actual (test-data/golden/stdio/write-stdout.z3.out) output:
                      F-expect>|output
                      F-actual>|[Crux] Found counterexample for verification goal
                      F-expect>|[Crux] Overall status: Valid.
                      F-actual>|[Crux]   test-data/golden/stdio/write-stdout.c:9:13: error: in main
                      F-actual>|[Crux]   Failed to load function handle
                      F-actual>|[Crux]   Details:
                      F-actual>|[Crux]     No implementation or override found for pointer: "\SOH_write"
                      F-actual>|[Crux]     The given pointer could not be resolved to a callable function
                      F-actual>|[Crux]     No implementation or override found for pointer: "\SOH_write"
                      F-actual>|[Crux]     Attempting to load callable function with type: i64(i32, i8*, i64)
                      F-actual>|[Crux]       Via pointer: Global symbol "\SOH_write" (1, 0x0:[64])
                      F-actual>|[Crux]     In memory state:
                      F-actual>|[Crux]       Stack frame main
                      F-actual>|[Crux]         No writes or allocations
                      F-actual>|[Crux]       Base memory
                      F-actual>|[Crux]         Allocations:
                      F-actual>|[Crux]           GlobalAlloc 6 0x26:[64] Immutable 1-byte-aligned [global variable  ] .str.1
                      F-actual>|[Crux]           GlobalAlloc 5 0x8:[64] Immutable 1-byte-aligned [global variable  ] .str
                      F-actual>|[Crux]           GlobalAlloc 4 0x0:[64] Immutable 1-byte-aligned [defined function ] main
                      F-actual>|[Crux]           GlobalAlloc 3 0x0:[64] Immutable 1-byte-aligned [external function] crucible_assert
                      F-actual>|[Crux]           GlobalAlloc 2 0x0:[64] Immutable 1-byte-aligned [external function] llvm.dbg.value
                      F-actual>|[Crux]           GlobalAlloc 1 0x0:[64] Immutable 1-byte-aligned [external function] _write
                      F-actual>|[Crux]         Writes:
                      F-actual>|[Crux]           Indexed chunk:
                      F-actual>|[Crux]             5 |->   *(5, 0x0:[64]) := "output\n\NUL"
                      F-actual>|[Crux]             6 |->
                      F-actual>|[Crux]               *(6, 0x0:[64]) := "test-data/golden/stdio/write-stdout.c\NUL"
                      F-actual>|[Crux] Overall status: Invalid.

                  Use -p '/write-stdout/&&/loop.clang11.z3 4.8.10.write-stdout.write-stdout crux output/' to rerun this test only.
```

**Comment by RyanGlScott** [comment]
> I still am not sure why the CI runners have not encountered this issue.

I was also baffled by this, and some further digging reveals the sad truth: the CI is inadvertently skipping over the `crux-llvm` test suite on macOS. See #999.

Some local debugging reveals that this issue happens pretty consistently across a range of different LLVM versions of bindists, both those obtained from Homebrew and those obtained from the LLVM downloads page. I still don't have any idea why the name mangling is happening in the bitcode of all places, so I've posted a question to the LLVM Discord in the hopes that someone knowledgeable will chime in.

**Comment by RyanGlScott** [comment]
The folks on the LLVM Discord reaffirmed the conclusion in https://github.com/GaloisInc/crucible/issues/885#issuecomment-942807148: namely, that because macOS's `.h` files are using `__asm` to specify a precise name to use, that name will inevitably appear in the bitcode that Clang produces. In other words, we don't have much of an alternative here besides manually stripping off the `\01_` prefix that appears in each name. (We wouldn't be the first tool to use such a hack—see also HikariObfuscator/Hikari@fec77b9a4496cfd7c2b944cb7d8884482b726c8d, the fix for HikariObfuscator/Hikari#1.)

If we really want to be thorough here, we may also need to account for _suffixes_ that macOS attaches to function names under certain conditions. Here is how `__DARWIN_ALIAS` and friends are defined on my SDK:

<details>

```c
#define __DARWIN_ALIAS(sym)             __asm("_" __STRING(sym) __DARWIN_SUF_UNIX03)
#define __DARWIN_ALIAS_C(sym)           __asm("_" __STRING(sym) __DARWIN_SUF_NON_CANCELABLE __DARWIN_SUF_UNIX03)
#define __DARWIN_ALIAS_I(sym)           __asm("_" __STRING(sym) __DARWIN_SUF_64_BIT_INO_T __DARWIN_SUF_UNIX03)
#define __DARWIN_NOCANCEL(sym)          __asm("_" __STRING(sym) __DARWIN_SUF_NON_CANCELABLE)
#define __DARWIN_INODE64(sym)           __asm("_" __STRING(sym) __DARWIN_SUF_64_BIT_INO_T)

#define __DARWIN_1050(sym)              __asm("_" __STRING(sym) __DARWIN_SUF_1050)
#define __DARWIN_1050ALIAS(sym)         __asm("_" __STRING(sym) __DARWIN_SUF_1050 __DARWIN_SUF_UNIX03)
#define __DARWIN_1050ALIAS_C(sym)       __asm("_" __STRING(sym) __DARWIN_SUF_1050 __DARWIN_SUF_NON_CANCELABLE __DARWIN_SUF_UNIX03    )
#define __DARWIN_1050ALIAS_I(sym)       __asm("_" __STRING(sym) __DARWIN_SUF_1050 __DARWIN_SUF_64_BIT_INO_T __DARWIN_SUF_UNIX03)
#define __DARWIN_1050INODE64(sym)       __asm("_" __STRING(sym) __DARWIN_SUF_1050 __DARWIN_SUF_64_BIT_INO_T)

#define __DARWIN_EXTSN(sym)             __asm("_" __STRING(sym) __DARWIN_SUF_EXTSN)
#define __DARWIN_EXTSN_C(sym)           __asm("_" __STRING(sym) __DARWIN_SUF_EXTSN __DARWIN_SUF_NON_CANCELABLE)
```
</details>

In addition to the leading underscore, there are other macros used as suffixes. Worse yet, some of these suffixes are only defined if the user has defined certain macros:

<details>

```c
/*
 * symbol suffixes used for symbol versioning
 */
#if __DARWIN_UNIX03
#  if __DARWIN_ONLY_UNIX_CONFORMANCE
#    define __DARWIN_SUF_UNIX03         /* nothing */
#  else /* !__DARWIN_ONLY_UNIX_CONFORMANCE */
#    define __DARWIN_SUF_UNIX03         "$UNIX2003"
#  endif /* __DARWIN_ONLY_UNIX_CONFORMANCE */

#  if __DARWIN_64_BIT_INO_T
#    if __DARWIN_ONLY_64_BIT_INO_T
#      define __DARWIN_SUF_64_BIT_INO_T /* nothing */
#    else /* !__DARWIN_ONLY_64_BIT_INO_T */
#      define __DARWIN_SUF_64_BIT_INO_T "$INODE64"
#    endif /* __DARWIN_ONLY_64_BIT_INO_T */
#  else /* !__DARWIN_64_BIT_INO_T */
#    define __DARWIN_SUF_64_BIT_INO_T   /* nothing */
#  endif /* __DARWIN_64_BIT_INO_T */

#  if __DARWIN_VERS_1050
#    if __DARWIN_ONLY_VERS_1050
#      define __DARWIN_SUF_1050         /* nothing */
#    else /* !__DARWIN_ONLY_VERS_1050 */
#      define __DARWIN_SUF_1050         "$1050"
#    endif /* __DARWIN_ONLY_VERS_1050 */
#  else /* !__DARWIN_VERS_1050 */
#    define __DARWIN_SUF_1050           /* nothing */
#  endif /* __DARWIN_VERS_1050 */

#  if __DARWIN_NON_CANCELABLE
#    define __DARWIN_SUF_NON_CANCELABLE "$NOCANCEL"
#  else /* !__DARWIN_NON_CANCELABLE */
#    define __DARWIN_SUF_NON_CANCELABLE /* nothing */
#  endif /* __DARWIN_NON_CANCELABLE */

#else /* !__DARWIN_UNIX03 */
#  define __DARWIN_SUF_UNIX03           /* nothing */
#  define __DARWIN_SUF_64_BIT_INO_T     /* nothing */
#  define __DARWIN_SUF_NON_CANCELABLE   /* nothing */
#  define __DARWIN_SUF_1050             /* nothing */
#endif /* __DARWIN_UNIX03 */

#define __DARWIN_SUF_EXTSN              "$DARWIN_EXTSN"
```
</details>

It's a corner case, but it would probably be worth stripping off the `$UNIX2003`, `$INODE64`, `$1050`, `$NOCANCEL`, and `$DARWIN_EXTSN` suffixes just to be on the safe side.
